macrosynergy/panel/adjust_weights.py

```diff
@@ -0,0 +1,258 @@
+    all_missing = zns_missing_in_weights + weights_missing_in_zns
+    if all_missing:
+        raise ValueError(f"Missing tickers: {all_missing}")
+
+    return df_weights_wide, df_adj_zns_wide
+
+
+def normalize_weights(df_weights_wide: pd.DataFrame) -> pd.DataFrame:
+    """
+    Weights are normalized by dividing each row by the sum of the row. Function exists to
+    allow easy modification of normalization method.
+
+    Parameters
+    ----------
+    df_weights_wide : pd.DataFrame
+        DataFrame with weights in wide format. (one column per cid)
+
+    Returns
+    -------
+    pd.DataFrame
+        DataFrame with normalized weights (sum of each row is 1).
+    """
+    df_weights_wide = df_weights_wide.div(
+        df_weights_wide.sum(axis="columns"), axis="index"
+    )
+
+    norm_rows = df_weights_wide.sum(axis="columns").apply(lambda x: np.isclose(x, 1))
+    all_nan_rows = df_weights_wide.index[df_weights_wide.isnull().all(axis="columns")]
+
+    # assert that all rows sum to 1 or are all NaN
+    if not norm_rows.all() and all_nan_rows.size == 0:
+        raise Exception("Normalization failed; weights do not sum to 1")
+
+    return df_weights_wide
+
+
+def adjust_weights(
+    df: QuantamentalDataFrame,
+    weights: str,
+    adj_zns: str,
+    method: Callable[[Number], Number],
+    params: Dict[str, Any] = {},
+    cids: List[str] = None,
+    adj_name: str = "ADJWGT",
+):
+    """
+    Adjusts the weights of a given xcat by a given adjustment xcat using a given method.
+
+    Parameters
+    ----------
+    df : QuantamentalDataFrame
+        QuantamentalDataFrame with weights and adjustment xcats for all cids.
+    weights : str
+        Name of the xcat containing the weights.
+    adj_zns : str
+        Name of the xcat containing the adjustment factors.
+    method : Callable
+        Function that will be applied to the weights to adjust them. This function must
+        conform to `f(x: Number, *args) -> Number`.
+    params : Dict[str, Any], optional
+        Parameters to be passed to the method function. Default is {}.
+    cids : List[str], optional
+        List of cids to adjust. If None, all cids will be adjusted. Default is None.
+    adj_name : str, optional
+        Name of the resulting xcat. Default is "ADJWGT".
+    """
+
+    if not isinstance(df, QuantamentalDataFrame):
+        raise TypeError("df must be a QuantamentalDataFrame")
+
+    df = QuantamentalDataFrame(df)
+    result_as_categorical = df.InitializedAsCategorical
+
+    if cids is None:
+        cids = df["cid"].unique().tolist()
+
+    check_types(weights, adj_zns, method, params, cids)
+
+    df, r_xcats, r_cids = reduce_df(
+        df, cids=cids, xcats=[weights, adj_zns], intersect=True, out_all=True
+    )
+
+    check_missing_cids_xcats(weights, adj_zns, cids, r_xcats, r_cids)
+
+    df_weights_wide, df_adj_zns_wide = split_weights_adj_zns(df, weights, adj_zns)
+
+    df_weights_wide = normalize_weights(df_weights_wide)
```

Comment:
We are we normalizing ahead of the adjustment? I can understand the normalisation below (line 232), but isn't this one redundant? It is just a scaler applied.

Reply:
removed